[dickongwd] iP

dickongwd opened the pull request:
# Duke

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/dickongwd/ip/releases/tag/A-Jar).
2. run it using `java -jar duke.jar`
3. add your tasks
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        // code here
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

dickongwd committed:
Change chatbot name and implement initial greet and exit

dickongwd committed:
Add echo functionality

dickongwd committed:
Add list functionality

dickongwd committed:
Add Task class, Command enum and mark/unmark functionality

dickongwd committed:
Add Todo, Event and Deadline classes

dickongwd committed:
Add functionality for todo, deadline and event

dickongwd committed:
Add testing using I/O redirection

dickongwd committed:
Handle input errors

dickongwd committed:
Update tests to account for new error handling

dickongwd committed:
Add delete functionality

dickongwd committed:
Add ui package

dickongwd committed:
Add save functionality with storage package

dickongwd committed:
Add storage package

dickongwd committed:
Add save functionality

dickongwd committed:
Add command, parser and common packages

dickongwd committed:
Add duke storage file to gitignore

dickongwd committed:
Refactor decode to use parser

dickongwd committed:
Refactor by separating into packages

dickongwd committed:
Fix missing imports

dickongwd committed:
Add documentation

dickongwd committed:
Allow Deadline to handle dates

dickongwd committed:
Allow Event to handle dates

dickongwd committed:
Display dates in different format

dickongwd committed:
Handle dates using date objects

dickongwd committed:
Merge branch 'add-gradle-support'

dickongwd committed:
Change mainClass for gradle config

WinstonLeonard commented on src/main/java/duke/common/Messages.java:
Excellent design choice by making all the repeated messages static final

WinstonLeonard reviewed (commented):
Great job of implementing the Duke Chatbot. OOP principles can be clearly seen in your code. All of the code seems to be following the Java styling guide. Keep up the good work!

dickongwd committed:
Add tests for parsing events

dickongwd committed:
Add tests for command type parsing

dickongwd committed:
Add tests for execution of deadline and invalid commands

dickongwd committed:
Add JUnit tests

dickongwd committed:
Add missing @throws for javadoc

dickongwd committed:
Add documentation for various command functions

dickongwd committed:
Refactor marked to sound like a boolean variable

dickongwd committed:
Add functionality to search for tasks by keyword

dickongwd committed:
Update text ui tests for find

dickongwd committed:
Merge branch 'branch-level-9'

dickongwd committed:
Merge branch 'branch-A-CodingStandard'

dickongwd committed:
Merge branch 'branch-A-JavaDoc'

vivienherq commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * The main class for the Duke chatbot.
+ */
+public class Duke {
```
LGTM!

kokrui commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+        String command = matcher.group("command").trim();
+        String args = matcher.group("arguments").trim();
+
+        switch (command) {
```
Based on the SE-EDU Coding Standard, a Basic Coding Standard rule the `case`s in a switch-case statement should be on the same indentation level as the `switch` statement

kokrui commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+            throw new DukeParseException("Event cannot be parsed");
+        }
+        String desc = matcher.group("desc").trim();
+        String from = matcher.group("from").trim();
```
It could be helpful for readability if the logical units within each code block are separated by one blank line, especially when methods are slightly more complex

kokrui commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Override
+    public String[] execute() {
```
It would be nice if this `execute` method had a JavaDoc comment like the corresponding methods in other subclasses of `Command`, especially since it's relatively complex

kokrui commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,76 @@
+        String mark = matcher.group("mark");
+        String input = matcher.group("taskString");
+        Task task;
+        switch (type) {
```
Based on the SE-EDU Coding Standard, a Basic Coding Standard rule the `case`s in a switch-case statement should be on the same indentation level as the `switch` statement

kokrui reviewed (commented):
LGTM! Minor nits on a few SE-EDU coding standard violations, with the only Basic violation not adhered to being the switch-case indentation levels! Otherwise, this is a really clean and comprehensive implementation!

dickongwd committed:
Setup basic GUI using javafx

dickongwd committed:
Configure javafx to use controllers and echo input

dickongwd committed:
Connect javafx GUI with chatbot response

dickongwd committed:
Move tasks into tasks package

dickongwd committed:
Fix decoder where regex matcher was not called

dickongwd committed:
Remove text ui class and tests

dickongwd committed:
Merge branch 'branch-Level-10'

dickongwd committed:
Setup checkstyle

dickongwd committed:
Refactor to follow checkstyle guidelines

bobscodedump commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,92 @@
+     * @return the task retrieved after decoding
+     */
+    public static Task decode(String encodedTask) {
+        Pattern pattern = Pattern.compile("^(?<type>.)\\|(?<mark>.)\\|(?<taskString>.+)$");
```
Perhaps you could shorten this a bit and abstract out the code as it crosses the 30 line threshold

bobscodedump commented on src/main/java/duke/common/Messages.java:
```diff
@@ -0,0 +1,11 @@
+ * A list of common messages used in the chatbot.
+ */
+public class Messages {
+    public static final String MESSAGE_WELCOME = "Hello! I'm Buzu.";
```
Good usage of abstraction of messages

bobscodedump reviewed (commented):
Overall great code! You used abstraction well and organised your code in a way that's easy to comprehend

dickongwd committed:
Enable assertions when running with gradle

dickongwd committed:
Add assertions for valid task numbers

Invalid task numbers are currently being handled via IndexOutOfBounds
exceptions.

We should also add asserts to document key assumptions that
should hold at various points in the program.

Let's add assertion statements for commands that use these task
numbers so that it is clear that the task numbers should be valid.

dickongwd committed:
Improve code quality

dickongwd committed:
Merge pull request #2 from dickongwd/branch-assertions

Add assertions

dickongwd committed:
Merge pull request #3 from dickongwd/branch-code-quality

Improve code quality

dickongwd committed:
Add functionality to specify data source

dickongwd committed:
Refactor to have getters and setters for Duke objects

dickongwd committed:
Add functionality for loading other data files

Duke now only loads from the default data file: .duke-storage.txt

Users of the chatbot may want to load from a specific data file, and
this can be used for importing/exporting saved tasks.

Let's refactor commands to accept the Duke object instead to provide
context so the load command can overwrite Duke's storage object.

dickongwd committed:
Fix assertion statements for task numbers

dickongwd committed:
Fix failing test case

dickongwd committed:
Fix checkstyle errors

dickongwd committed:
Merge branch 'branch-flexible-data-source'

dickongwd committed:
Add product screenshot

dickongwd committed:
Add user guide

dickongwd committed:
Fix find command

dickongwd committed:
Update user guide

dickongwd committed:
Update README with link to jar file
